src/Workspaces/Core/Portable/Diagnostics/DiagnosticData.cs

```diff
@@ -336,13 +349,28 @@ public static DiagnosticData Create(Diagnostic diagnostic, Document document)
                                                 .WhereNotNull()
                                                 .ToReadOnlyCollection();
 
+            var additionalProperties = GetAdditionalProperties(document, diagnostic);
+
+            var documentPropertiesService = document.Services.GetService<DocumentPropertiesService>();
+            var diagnosticsLspClientName = documentPropertiesService?.DiagnosticsLspClientName;
+
+            if (diagnosticsLspClientName != null)
+            {
+                if (additionalProperties == null)
+                {
+                    additionalProperties = ImmutableDictionary.Create<string, string>();
+                }
+
+                additionalProperties = additionalProperties.Add(nameof(documentPropertiesService.DiagnosticsLspClientName), diagnosticsLspClientName);
+            }
+
             return Create(diagnostic,
                 project.Id,
                 project.Language,
                 project.Solution.Options,
                 location,
                 additionalLocations,
-                GetAdditionalProperties(document, diagnostic));
+                additionalProperties);
```

Comment:
i'm confused why this needs to be in the diagnostic.  if we can already ask the Document for it's `DiagnosticsLspClientName`, then why can't people just look at that instead of the diagnostic itself?

Reply:
So the DiagnosticData doesn't hold onto the Document or DocumentId as far as I can tell; so if we didn't capture the value here it's not easy for other layers (like the error list) to re-filter -- once you're past this point you don't know which document it came from. And since file names can be remapped, that would make it hard to re-find the document as best I can tell.

It's possible I'm missing something else here but it seemed like GetAdditionalProperties was "where" we did the last chance ability to capture from the Document because beyond here it was "too late".